fixes #1908 enhancement request

v3nom opened the pull request:
Extended amd-dependency comment with name attribute, which is used to generate more human-readable JavaScript AMD output. It also simplifies JavaScript AMD module use in TS code. #1908

msftclas commented:
Hi __@v3nom__, I'm your friendly neighborhood Microsoft Pull Request Bot (You can call me MSBOT). Thanks for your contribution!
    <span>
        In order for us to evaluate and accept your PR, we ask that you sign a contribution license agreement. It's all electronic and will take just minutes. I promise there's no faxing. https://cla.microsoft.com.
    </span>

TTYL, MSBOT;

msftclas commented:
__@v3nom__, Thanks for signing the contribution license agreement so quickly! Actual humans will now validate the agreement and then evaluate the PR.
<br />Thanks, MSBOT;

DickvdBrink commented:
Hmz, weird, some files are now changed to 755. You might want to look at this: [link](http://stackoverflow.com/questions/1257592/how-do-i-remove-files-saying-old-mode-100755-new-mode-100644-from-unstaged-cha)

yuit commented on src/compiler/parser.ts:
```diff
@@ -4680,7 +4680,7 @@ module ts {
             var triviaScanner = createScanner(languageVersion, /*skipTrivia*/false, sourceText);
             var referencedFiles: FileReference[] = [];
-            var amdDependencies: string[] = [];
+            var amdDependencies: {path:string; name:string}[] = [];
```
Minor: Please put space between ```:```and ```string```

yuit commented:
Please add more tests

DanielRosenwasser commented on src/compiler/parser.ts:
```diff
@@ -4720,10 +4720,16 @@ module ts {
-                    var amdDependencyRegEx = /^\/\/\/\s*<amd-dependency\s+path\s*=\s*('|")(.+?)\1/gim;
+                    var amdDependencyRegEx = /^\/\/\/\s*<amd-dependency/gim;
+                    var pathRegex = /path\s*=\s*('|")(.+?)\1/gim
+                    var nameRegex = /name\s*=\s*('|")(.+?)\1/gim
```
Semicolons.

DanielRosenwasser commented on src/compiler/parser.ts:
```diff
@@ -4720,10 +4720,16 @@ module ts {
-                        amdDependencies.push(amdDependencyMatchResult[2]);
+                        var pathMatchResult = pathRegex.exec(comment)
+                        var nameMatchResult = nameRegex.exec(comment)
+                        if(pathMatchResult){
```
Space after `if`, space after parens

DanielRosenwasser commented on src/compiler/parser.ts:
```diff
@@ -4720,10 +4720,16 @@ module ts {
+                        var pathMatchResult = pathRegex.exec(comment)
+                        var nameMatchResult = nameRegex.exec(comment)
+                        if(pathMatchResult){
+                            amdDependencies.push({path:pathMatchResult[2], name:nameMatchResult?nameMatchResult[2]:null});
```
Space after `{`, before `}`, before/after `?` and `:`, use `undefined` instead of `null`.

mhegazy commented on src/compiler/parser.ts:
```diff
@@ -4720,10 +4720,16 @@ module ts {
                     }
 
-                    var amdDependencyRegEx = /^\/\/\/\s*<amd-dependency\s+path\s*=\s*('|")(.+?)\1/gim;
+                    var amdDependencyRegEx = /^\/\/\/\s*<amd-dependency/gim;
```
there should be a space after "amd-dependency" and before path or name. i would add \s+ back to the regex

mhegazy commented:
For the tests, you can look at [tests\cases\compiler\amdDependencyComment1.ts](https://github.com/Microsoft/TypeScript/blob/master/tests/cases/compiler/amdDependencyComment1.ts) as a reference.

v3nom commented:
Thanks for comments. I have updated code to match code style. Added /s to regexp and added test cases following existing amdDependencyComment(1/2).ts example

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3925,6 +3925,12 @@ module ts {
                     write(", ");
                     emit(imp.name);
                 });
+                forEach(node.amdDependencies, amdDependency => {
```
what if you have multiple dependencies, ones with names and ones without. this will get the order our of sync, since name is optional and path is not. e.g.:
```TypeScript
///<amd-dependency path='module1' />
///<amd-dependency path='module2' name='m2'/>
```

will emit:
```JavaScript
define(["require", "exports", "module1", "module2"], function (require, exports, m2) { }
```

which is not what you want. so either you reorder them, so that ones with names come first, which we do not do usually as we like to keep emitted code as close as possible to generated code. or you give the other ones unused random names consider using createTempVariable utility in the emitter.

Also worth adding a new test for this case.

v3nom replied:
Giving random names would add unnecessary garbage to emitted code. Since import order does not matter for AMD modules, reordering would make most sense.

Changing order is a bit intrusive to original code, but amd-dependency comment is not a language feature, just a utility. Its also not documented in official materials, so I think violating some project principles is justifiable.

mhegazy commented on src/compiler/parser.ts:
```diff
@@ -4710,10 +4710,22 @@ module ts {
+                            var amdDependency = {path: pathMatchResult[2], name: nameMatchResult ? nameMatchResult[2] : undefined };
+                            // AMD dependencies with names have to go first in define header
+                            if (nameMatchResult) {
+                                amdDependencies.unshift(amdDependency);
```
why not just sort the list in the emitter? i do not think we will have that many AMD imports that a sort would have a  noticeable perf impact. this way the parser always have the truth, and the emitter can chose to do some extra work to ensure a valid transformation.

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3902,11 +3902,25 @@ module ts {
+            function sortAMDModules(amdModules: {name: string; path: string}[]) {
+                // AMD modules with declared variable names goes first
+                return amdModules.sort((moduleA, moduleB) => {
+                    if (moduleA.name == moduleB.name) {
```
nit: "===" instead of "=="

mhegazy commented on src/compiler/emitter.ts:
```diff
@@ -3902,11 +3902,25 @@ module ts {
+                return amdModules.sort((moduleA, moduleB) => {
+                    if (moduleA.name == moduleB.name) {
+                        return 0;
+                    } else if (moduleA.name == undefined) {
```
nit: `!moduleA.name` instead of `moduleA.name == undefined` as an empty name is still an invalid name. an later on you do check for amdDependency.name.

mhegazy commented:
Thanks @v3nom. I think we are good to go. just a couple of comments, and if you can merge master into your branch as well. thanks again.

mhegazy commented:
thanks!
